### PR TITLE
Adapt contributor setup instructions for Windows. Fixup for lack of pytest-django dependency on pytest.

### PR DESCRIPTION
**Pull request by ptressel**
Changes to contribute.rst, so that the setup can be run on Windows, in addition to *ix systems.

- Remove leading "$" prompt and trailing comments on commands, so they can be copy-pasted.
- Move descriptions of what each command will do out of *ix-style comments (which are not Windows comments) to preceding text.
- Separate creation of virtualenv from installation of packages so that the above descriptions match what the commands are doing, and so that users can do each step separately.

Also, in a change to the same file:

- Add distinction between use of password or SSH keys to authenticate to GitHub, as the commands as given will not work for people who are using SSH keys.

Changes to Pipfile, Pipfile.lock:

The requirements.txt for the pytest-django package lacks pytest as a dependency.  The Pipfile generated from our own requirements.txt happened to insert pytest-django before pytest, so pytest-django was installed without pytest.  Apparently this meant pytest-django was unable to insert what it needed into pytest.  As a result, pytest could not find the test_project settings, and also lacked the --ds option inserted by pytest-django.

Note you won't see this problem unless you are starting from a new, empty virtualenv, that does not already have pytest.

As a temporary fix, I've moved pytest prior to pytest-django in the Pipfile.  The real fix would be to get the pytest-django folks to add pytest to their requirements.txt.  If I can figure out what the minimum pytest version is for pytest-django, I'll send them a PR, else will open an issue.

When pipenv generated Pipfile.lock from the new Pipfile, it included package colorama.  This is a dependency for Windows -- it allows coloring text in a Windows terminal window.  It won't affect other systems.

https://pypi.python.org/pypi/colorama

### CODE REVIEW

**Comment by ariannedee** [comment]
@ptressel, do you have a particular reason to include separate instructions for github users using different authentication methods? IMO, the user should know how to use git clone on their environment or look it up if they are running into problems (I had to keep looking it up when I had 2-FA on). I think that it bloats our documentation unnecessarily, but am open to other opinions. @melinath?

**Comment by ariannedee** [comment]
Also, @ptressel, Stephen has submitted a PR to move the docs requirements out of the Pipfile #92, because of #91, so your changes to the docs section will have to be updated after that gets merged.

File: docs/contribute.rst [hunk 1]
```diff
@@ -13,28 +13,79 @@ Start Developing
...
+
+    git clone https://github.com/your-username/django-graph-api.git
+
+The SSH keys version will look like:
+
+::
+
+    git clone git@github.com:your-username/django-graph-api.git
+
+After cloning, navigate into the new directory.
```
Comment (melinath) [inline docs/contribute.rst]: I agree with @ariannedee that this detail is more than we want in the docs. If we're concerned about people not knowing how to fork or clone a repository using github, we should be directing them to Github docs on the subject. (But it would be good to have that as a separate PR.)

**Comment by ptressel** [comment]
I'm confused by this:

> This PR touches a lot of the docs that are unrelated to Windows setup though.

There are two commits.  One commit has the Windows changes, and only alters a single file.  The other commit is the workaround for the lack of pytest as a dependency for pytest-django.  The reason it was added to this PR is that the setup failed without it, and I diagnosed the cause after the doc changes were complete and in a commit.

Do you want that as a separate PR?  If it is split out, then it should be first, since it affects whether the setup works at all.  Note it doesn't just affect the setup modified for Windows -- the original instructions fail as well, unless one happens to have pytest already installed, and after the first (failed) run, it will be.

I'm also confused by this:

> but ideally each of the bullet points in the description should be in their own PR (or max 2).

In all other open source projects I've worked on, they wanted the opposite.  They wanted everything relating to a coordinated change together, not just in a single PR, but in a single commit.  If it is the *description of the PR* that is the objection, then I could remove the explanation, and just say, "Make it work on all of Windows, MacOS, and Linux."  But I figured that the explanation would make it more understandable, and less likely to get pushback from folks who do not use Windows, and thus don't immediately recognize what won't work.

So, please bear with me, because this objection is very surprising.  It seems like a *lot* more work for everyone to make a single change in each of a half dozen PRs, especially since none of those but the last will work for the stated purpose, i.e. allowing a Windows user to do the setup.  If someone sent me a bunch of separate PRs, none of which had working code / docs except the last one, I'd say, please combine these.  So...are you really sure you want that?  This is only a few lines of changes...

> do you have a particular reason to include separate instructions for github users using different authentication methods?

Yes, it was because the original statement was absolute, and had no indication that it was incorrect for some users.  If it is modified to make it clear that it is only an example, and that people will have to change it depending on their authentication, then that will avoid the failure if people (like me) who use SSH keys were to copy-paste that line.  I didn't get caught, and accidentally copy-paste that line, but I teach git / GitHub use...

I'm assuming this would be preferred in a separate PR.  The PR would have maybe half a line of text in it...

If you want to make the level of detail on git / GitHub usage consistent, then maybe none of that should be in there.  But that may lead to more questions from new users, for whom git / GitHub is a new thing.

In other projects, we've gradually moved to being more and more explicit with instructions.  Folks got tired of explaining the same thing over and over, so it got added in detail to the docs.

Fortunately, this isn't time critical.  Unfortunately, I have other things that I really have to get done first.  That may mean there isn't action on this for a while, unless I can sneak it in.

**Comment by ptressel** [comment]
Here's another option:  I could see if the pytest-django folks are willing to fix their requirements.txt.  We could wait until they fix it.  That would allow removing the workaround.  However, we don't know how quickly they will respond or incorporate the fix.  And the workaround doesn't break anything.  It seemed like a cheap fix that permits the pytest-django negotiation to take as long as needed.

I updated to original PR description to split the Windows part off from the GitHub auth part.  The line item about having each command do one action is in the Windows section in part because it made the commands match the text descriptions that were taken from comments and merged into the existing text.  But it's also because it was nuking the virtualenv repeatedly during testing because the "--python 3.6" was attached to a command that normally is not destructive.  It's better to have safe, idempotent, commands separate from destructive commands, especially if the destructive command is going to cause a bunch of packages to be re-downloaded...  That's not strictly for Windows, so it could be split out.

**Comment by ariannedee** [comment]
@ptressel Thanks for the explanation.

Here are my thoughts related to each of the points in the description:

> Changes to contribute.rst, so that the setup can be run on Windows, in addition to *ix systems.

Sounds great. We definitely want this. (Including the Pipfile and Pipfile.lock changes)

> Remove leading "$" prompt and trailing comments on commands, so they can be copy-pasted.

I like the '$' and was trying to use that and 'venv$' (or something like that) to be able to indicate which commands could be run in the shell directly and which ones had to be in a virtual environment. The latter ones might have been removed at some time during the sprint, but that was the original idea.

> Move descriptions of what each command will do out of *ix-style comments (which are not Windows comments) to preceding text.

I didn't realize this was a Windows related change before it was edited. But also, it is more related to the expectation that you can copy and paste directly, instead of understanding what the commands do and why you do them.

> Separate creation of virtualenv from installation of packages so that the above descriptions match what the commands are doing, and so that users can do each step separately.

I had also written these as separate steps before and they were changed during the sprint. I think it is better for them to be separate, but it is also unrelated to Windows setup and I could see them being in another issue. I am fine with leaving this in.

> Add distinction between use of password or SSH keys to authenticate to GitHub, as the commands as given will not work for people who are using SSH keys.

I think that this makes our documentation a lot more verbose, and it would be better to link to github pages that explain how to clone a repo. I think it holds us up to much more difficult standards to be teaching Github and Pipenv in our docs, instead of focusing on DGA and providing links that allow them to learn what they need to learn in those respective places. If Github or Pipenv change, then we need to change our docs.

> In other projects, we've gradually moved to being more and more explicit with instructions. Folks got tired of explaining the same thing over and over, so it got added in detail to the docs.

I understand this, but am torn about whether or not now is the right time to be doing that. Once there are actually multiple people with a lot of the same questions, we will definitely update it, but I would prefer it to be an organic process. I worry about overfilling it with information that makes it harder to know what the important parts are that make setting up this project different from other projects.

Overall, I asked to break these up (not necessarily the best approach) mostly because I think there is a lot to discuss and I'm not sure the best way of doing that. This is my first open source project, and I am used to being able to discuss things in real-time with coworkers and thought it would be easier to discuss each item if they were in separate issues. What seem like (and may be) small changes, feel quite opinionated to me about how we should be approaching our documentation, which feels like a larger discussion. It is also easy to miss things in large github comments.

Going forward, if you removed the github stuff, I'd be happy to merge it. If you really think it should be in there, please make another issue and we can talk there about the best approach to being beginner friendly. I would prefer links to external sources that are maintained by them. If you want to discuss anything directly, feel free to message me on Slack. And thanks for all the trouble shooting getting it to work on Windows! I have tried to get Python on Windows working for a friend and found it to be a big pain.

**Comment by ptressel** [comment]
Here's another option:  I could see if the pytest-django folks are willing to fix their requirements.txt.  We could wait until they fix it.  That would allow removing the workaround.  However, we don't know how quickly they will respond or incorporate the fix.  And the workaround doesn't break anything.  It seemed like a cheap fix that permits the pytest-django negotiation to take as long as needed.

I updated to original PR description to split the Windows part off from the GitHub auth part.  The line item about having each command do one action is in the Windows section in part because it made the commands match the text descriptions that were taken from comments and merged into the existing text.  But it's also because it was nuking the virtualenv repeatedly during testing because the "--python 3.6" was attached to a command that normally is not destructive.  It's better to have safe, idempotent, commands separate from destructive commands, especially if the destructive command is going to cause a bunch of packages to be re-downloaded...  That's not strictly for Windows, so it could be split out.

(Not sure what happened to this comment -- I posted it last night, but now my browser has it open in the comment editor.  I'm going to post it again -- if it's a duplicate, I'll delete it.  Maybe this is due to my flaky network connection...)

**Comment by ptressel** [comment]
Ok, I have about 2 hours with a good network connection, so let me refactor this.

I'll split out the Pipfile and Pipfile.lock changes into a separate PR, because those two are tied together, *in spite of the fact* that the auto-generated Pipfile.lock has a Windows-required change:  That was added automatically by pipenv.  Again, I changed *only* Pipfile, and then Pipfile.lock was produced by pipenv.

I'll remove the edits that show both password and SSH keys forms of auth.  I recommend either showing both, or stating that the shown password form is not normative, or leaving in the description of how to get the repo path, or removing the entire command with the password form, and referring to GitHub docs.  That can be a separate issue, later.

At that point, the remaining edits are Windows related.  Let me explain a bit more about how things work on Windows, which I hope will clear things up.

>> Remove leading "$" prompt and trailing comments on commands, so they can be copy-pasted.

> I like the '$' and was trying to use that and 'venv$' (or something like that) to be able to indicate which commands could be run in the shell directly and which ones had to be in a virtual environment. The latter ones might have been removed at some time during the sprint, but that was the original idea.

Here's what the prompt looks like in Windows, using the standard cmd, when I have the current folder (not directory) set to my repo.

D:\PTDocuments\python\django-graph-api>

The only difference in Windows PowerShell is that there's a PS and a space before the path prompt.  I don't recall what the git bash prompt looks like -- it damaged my registry, and I won't run it and don't recommend it.  In any case, it's not a native Windows thing.

It looks exactly the same after I activate the virtual environment.  The dollar sign is simply not something one sees in Windows.

I like the concept of indicating which commands are in the virtualenv, but don't think text in the command line is the best way to show it.  Would be better to use some other indicator that doesn't look like part of the command, e.g. an indicator *outside* of the text box enclosing the command, or different color backgrounds in the text box, or...  In any case, it should be explained, not left for people to guess what the meaning is.  Doing CSS-like adjustments could be tricky in Markdown, so I'd rather not do anything about it in this PR.  I'm guessing that would not be wanted anyway.  ;-)

>> Move descriptions of what each command will do out of *ix-style comments (which are not Windows comments) to preceding text.

> I didn't realize this was a Windows related change before it was edited. But also, it is more related to the expectation that you can copy and paste directly, instead of understanding what the commands do and why you do them.

This really is a Windows thing.  Windows commands *do not have trailing comments*.  This is also true for a Windows "batch file", written in the Windows standard scripting language.  Comment *go on a separate line* and are prefixed by the text REM (for remark).  A comment thus looks like:

REM This is a comment.

To a Windows user, typically, everything on the command line is part of the command.

Re. copy-paste, we talked about that at the hackathon, and people liked the idea.  It was a secondary benefit that I mentioned, not the primary reason for the change.

>> Separate creation of virtualenv from installation of packages so that the above descriptions match what the commands are doing, and so that users can do each step separately.

> I had also written these as separate steps before and they were changed during the sprint.

I was surprised to see that change appear, and surprised that it was accepted.  It's true this isn't specific to Windows, but it's a Bad Idea for any OS.  People will not expect an innocent command that *installs packages* to blow away their environment.  I'm a big fan of fencing off dangerous commands from safe ones, and including warnings.  IMO it would have been better had the pipenv folks not allowed the exitstence of "--python 3.6" on the command line to mean tear down the environment -- I'd have preferred that they have an explicit command, like "pipenv create", that makes, or re-makes, an environment.

Changing it back did have an element of being annoyed that it nuked my environment and I had to download the packages all over.  I've been living with a 5GB high speed data cap, so I'm really sensitive to unnecessary network traffic.  If you're ok with it, I'll leave it as part of this PR.  One can think of it as making the commands match the textual description of what's happening.  But really, it's reverting the unwanted change, and not forcing people to lose their environment just to install or update packages.

Ok, I have about 1.5 hours of reliable network, so I'd better get on it...

**Comment by ptressel** [comment]
Ok, I've split this into two PRs, and reverted the SSH keys bits.  Here is the other PR, for Pipfile:

https://github.com/melinath/django-graph-api/pull/95

I'd recommend merging that one first, since tests won't run without it if one is installing from scratch.  Then, if you like having a linear commit history, I can rebase this one.

**Comment by ariannedee** [comment]
@ptressel Looks great, thanks so much for making those changes. I think it is a good medium between too little and too much information.

**Comment by ptressel** [comment]
I forgot to take the Pipfile stuff out of the PR description (derp...), so just did that, plus included some info from the discussion.
